### PR TITLE
Main feature for Touch Bar

### LINKED ISSUE #3350: Mac Touch Bar - Touch Bar Button logic, Version 2
### 💻 Environment

Platform: macOS

### 📒 Description

image

There are 3 different states:

**Passive state**
If the setting is enabled only the global toggl touch bar button is visible (In Control Strip). It has the same design as the menubar icon. (running and stopped state)

**Active state - timer running**
User has pressed the global Toggl touch bar button in Control strip. This updates the App region: 
- Adds a stop button right to the left of the global button. Button with stop icon `[ ■ ]`
- Adds info about current running entry to the left of stop button
- Adds a horizontally scrollable list of 5 latest unique time entries left of the stop button.
```
[ Item1 | Item2...] | [Running item info] | [ ■ ] | [Global Button]
```

**Active state - timer stopped**
User has pressed the global Toggl touch bar button in Control strip. This updates the App region:
- Adds a start button right to the left of the global button. Button with play icon `[ ▶ ]`
- Adds a horizontally scrollable list of 5 latest unique time entries left of the start button.
```
[ Item1 | Item2...] | [ ▶ ] | [Global Button]
```

_Pressing the global button again when the detailed view is opened will close the detailed view_

### LINKED ISSUE #3351: Mac Touch Bar - Get last 5 unique time entries. Version 2
### 💻 Environment

Platform: macOS

### 📒 Description
- Create a method that returns a list of 5 recent unique entries

These entries will be displayed in the touch bar. Specific logic in this issue: #3350

### PR DESCRIPTION
**Pull request by NghiaTranUIT**
### 📒 Description
This PR introduce all main logic of the Touch Bar

### 🕶️ Types of changes
**New feature** (non-breaking change which adds functionality)

### 🤯 List of changes
- [x] Get latest 5 Time Entry (https://github.com/toggl/toggldesktop/issues/3351)
- [x] Implement NSScrubber with custom layout -> Able to scrollable, the item size is depend on the text title
- [x] Binding the current running TE with TouchBar
- [x] Binding the Start button in TimerBar with Touch Bar
- [ ] Find the way to disable the default Touch Bar for the AutoCompleteTextField in Timer Bar (See https://github.com/toggl/toggldesktop/pull/3424#issuecomment-540995289)
- [x] Better approach to calculate the width size of TouchBar Item (from title length)

### 👫 Relationships
Closes #3351 
Closes #3350 

### 🔎 Review hints
<Update Later>

### CODE REVIEW

**Comment by NghiaTranUIT** [comment]
# Known-issues 
## 1. The Timer Bar TextField lost the default Touch Bar when we focus on Project Text Field in Editor (allowsEditingTextAttributes=true)


### How to reproduce
1. Open the app
2. Check that the Touch bar works well in Timer Bar and the rest of the app
3. Open Time Entry Editor and click on the Project Text Field
4. The Touch bar is replaced with the default macOS touch bar.
5. Go back to the main view  -> Timer Bar textfield doens't show our touch bar anymore???

### What's allowsEditingTextAttributes?
`allowsEditingTextAttributes` is a property that allow the user to edit on TextField, which has AttributedString (our Project Text field).
Screen Shot 2019-10-11 at 16 34 10
Screen Shot 2019-10-11 at 16 34 37

Currently, we're using in ProjectAutoCompleteTextField
https://github.com/toggl/toggldesktop/blob/0ee4fefd07a9fd9454671ad3cec47fe885c323dc/src/ui/osx/TogglDesktop/Feature/TimeEntryEditor/AutoComplete/Input/ProjectAutoCompleteTextField.swift#L32

## Attempts 
- Tried to override `makeTouchBar` for each TextField and return our global Touch Bar -> Doesn't work 💥 
- Tried to disable default touch bar of Project Text Field -> Doesn't work 💥 

# Play-around 
- Enable `allowsEditingTextAttributes` on the second text field

=> The Touch bar in the First Text field is lost when the second text field is the first responder.

- Disable `allowsEditingTextAttributes` on the second text field

=> All Touch bars are remained.

Download 
[TouchBar-Issues.zip](https://github.com/toggl/toggldesktop/files/3716725/TouchBar-Issues.zip)

**Comment by IndrekV** [comment]
Did you try `Disable allowsEditingTextAttributes on project text field` with our app?

**Comment by NghiaTranUIT** [comment]
if we disable the `allowsEditingTextAttributes` from Project AutoCompleteTextField, whenever we put the cursor into the text field, all Attributes are removed. See the screenshot



Therefore, the Touch Bar known-issue is resolved itself.

**Comment by IndrekV** [comment]
I've tested this again and I think it's ok to keep the logic as it is. When user had project text field focused it will show text editing buttons. Once you move away from the project field it resumes the correct logic.

Only thing that I think we could improve is the listing of 5 unique entries.  Currently it only show description and no reference to the project. Do you have any suggestions how to improve this.

I approve the things done in the current PR. Great job Nghia!

File: src/ui/windows/TogglDesktop/TogglDesktop/SingleInstanceManager.cs [hunk 1]
```diff
@@ -12,6 +12,7 @@ public SingleInstanceManager()
         }
 
         public event Action BeforeStartup;
+    public event Action BeforeStartup;
```
Comment (IndrekV) [inline src/ui/windows/TogglDesktop/TogglDesktop/SingleInstanceManager.cs]: Why is there two of the same thing here?

Comment (NghiaTranUIT) [reply src/ui/windows/TogglDesktop/TogglDesktop/SingleInstanceManager.cs]: I'm not sure what the cs file is edited. I will revert the change.

**Comment by IndrekV** [comment]
Any ideas about showing project somehow in the touchbar unique entries list.

Ideas:
- Show a dot with project color (good because does not take space)
- Make the entry background the color of the project (Should be simple to do but make it really colorful. I'm not sure it's a good look)
- Show project name text (bad as takes text space, maybe show color dot and a few first character of project name)

**Comment by NghiaTranUIT** [comment]
## Screenshot
> Make the entry background the color of the project (Should be simple to do but make it really colorful. I'm not sure it's a good look)

Screen Shot 2019-10-31 at 13 52 04

Currently, the running TE would have the background, which is the same with Project Color -> Thus, it's easier to distinguish the "Running TE" and the "TE List" in the Touch Bar.

If we set the background for all TE in the list -> It's really difficult to distinguish IMO

**Comment by NghiaTranUIT** [comment]
Let me do an experiment to make the Touch Bar button with two title (Description on the top, and project in the bottom)

**Comment by NghiaTranUIT** [comment]
## Running State
Screen Shot 2019-10-31 at 14 34 45

## Normal State
Screen Shot 2019-10-31 at 14 34 40

@IndrekV how is your thought ? 🌮

**Comment by NghiaTranUIT** [comment]
This PR is ready for the second review round ♻️ 



## Change logs from last review
- New design for touch bar item (see the GIF)
- Improve how we get unique 5 Time Entries from the list: Only get unique TEs, TEs in same group won't present in the List.

**Comment by IndrekV** [comment]
Wow, just wow. I did not even dream we could make the time entries look so good. You are magician Nghia. 



Great job.

**Pull request merged**